File: examples/python/webhook.py
"""Dona webhook signature verification (Standard Webhooks) — stdlib only.

Headers on every delivery: webhook-id · webhook-timestamp · webhook-signature.
  signature = space-delimited "v1,<base64 HMAC-SHA256(base64decode(secret after whsec_), id.ts.raw_body)>"
  (two entries for 24 h after a secret rotation — accept if ANY matches)

⚠ Verify the RAW request bytes (Flask: request.get_data(), Django: request.body, FastAPI:
  await request.body()). Re-serialised JSON never matches. Then dedupe on webhook-id for 24 h.

    try:
        verify_webhook(os.environ["DONA_WEBHOOK_SECRET"], request.headers, request.get_data())
    except WebhookVerificationError:
        return "", 401
    return "", 204          # acknowledge fast; process asynchronously
"""

import base64
import hashlib
import hmac
import time
from collections.abc import Mapping

TOLERANCE_SECONDS = 300


class WebhookVerificationError(Exception):
    pass


def _header(headers: Mapping[str, str], name: str) -> str:
    for k, v in headers.items():
        if k.lower() == name:
            return v
    return ""


def verify_webhook(secret: str, headers: Mapping[str, str], raw_body: bytes | str, now: int | None = None) -> None:
    """Raises WebhookVerificationError unless the delivery is authentic and fresh."""
    msg_id = _header(headers, "webhook-id")
    ts = _header(headers, "webhook-timestamp")
    signatures = _header(headers, "webhook-signature")
    if not (msg_id and ts and signatures):
        raise WebhookVerificationError("missing webhook-* headers")
    try:
        timestamp = int(ts)
    except ValueError as e:
        raise WebhookVerificationError("bad webhook-timestamp") from e
    now = int(time.time()) if now is None else now
    if abs(now - timestamp) > TOLERANCE_SECONDS:
        raise WebhookVerificationError("timestamp outside tolerance")

    key = base64.b64decode(secret.removeprefix("whsec_"))
    body = raw_body.encode("utf-8") if isinstance(raw_body, str) else raw_body
    expected = hmac.new(key, f"{msg_id}.{ts}.".encode() + body, hashlib.sha256).digest()
    for entry in signatures.split(" "):
        version, _, sig = entry.partition(",")
        if version != "v1" or not sig:
            continue
        try:
            got = base64.b64decode(sig, validate=True)
        except ValueError:
            continue
        if hmac.compare_digest(got, expected):
            return
    raise WebhookVerificationError("no matching signature")

File: examples/typescript/src/webhook.ts
// Dona webhook signature verification (Standard Webhooks) — Node, no dependencies.
//
// Headers on every delivery: webhook-id · webhook-timestamp · webhook-signature.
//   signature = space-delimited "v1,<base64 HMAC-SHA256(base64decode(secret after whsec_), id.ts.rawBody)>"
//   (two entries for 24 h after a secret rotation — accept if ANY matches)
//
// ⚠ Verify the RAW request bytes. Parsing the JSON and re-serialising it changes the bytes and the
//   signature will never match. Then dedupe on webhook-id (it is the event id) for 24 h.
//
//   http.createServer(async (req, res) => {
//     const body = Buffer.concat(await Array.fromAsync(req));
//     try { verifyWebhook(process.env.DONA_WEBHOOK_SECRET!, req.headers, body); }
//     catch { res.writeHead(401).end(); return; }
//     res.writeHead(204).end();          // acknowledge fast; process asynchronously
//   });
import { createHmac, timingSafeEqual } from "node:crypto";

export const TOLERANCE_SECONDS = 300;

export class WebhookVerificationError extends Error {}

type HeaderBag = Record<string, string | string[] | undefined> | Headers;

function header(h: HeaderBag, name: string): string {
  const v = h instanceof Headers ? h.get(name) : h[name] ?? h[name.toLowerCase()];
  return (Array.isArray(v) ? v[0] : v) ?? "";
}

/** Throws WebhookVerificationError unless the delivery is authentic and fresh. */
export function verifyWebhook(
  secret: string,
  headers: HeaderBag,
  rawBody: Uint8Array | string,
  nowSeconds: number = Math.floor(Date.now() / 1000),
): void {
  const id = header(headers, "webhook-id");
  const ts = header(headers, "webhook-timestamp");
  const signatures = header(headers, "webhook-signature");
  if (!id || !ts || !signatures) throw new WebhookVerificationError("missing webhook-* headers");

  const timestamp = Number(ts);
  if (!Number.isInteger(timestamp)) throw new WebhookVerificationError("bad webhook-timestamp");
  if (Math.abs(nowSeconds - timestamp) > TOLERANCE_SECONDS) throw new WebhookVerificationError("timestamp outside tolerance");

  const key = Buffer.from(secret.startsWith("whsec_") ? secret.slice(6) : secret, "base64");
  const body = typeof rawBody === "string" ? Buffer.from(rawBody, "utf8") : Buffer.from(rawBody);
  const expected = createHmac("sha256", key).update(`${id}.${ts}.`).update(body).digest();

  for (const entry of signatures.split(" ")) {
    const [version, sig] = entry.split(",", 2);
    if (version !== "v1" || !sig) continue;
    const got = Buffer.from(sig, "base64");
    if (got.length === expected.length && timingSafeEqual(got, expected)) return;
  }
  throw new WebhookVerificationError("no matching signature");
}
